util/stmtsummary/evicted.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stmtsummary
+
+import (
+	"container/list"
+	"time"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// stmtSummaryByDigestEvicted contents digests evicted from stmtSummaryByDigestMap
+type stmtSummaryByDigestEvicted struct {
+	// record evicted data in intervals
+	// latest history data is Back()
+	history *list.List
+}
+
+// element being stored in stmtSummaryByDigestEvicted
+type stmtSummaryByDigestEvictedElement struct {
+	// beginTime is the begin time of current interval
+	beginTime int64
+	// endTime is the end time of current interval
+	endTime int64
+	// *Kinds* of digest being evicted
+	digestKeyMap map[string]struct{}
+}
+
+// spawn a new pointer to stmtSummaryByDigestEvicted
+func newStmtSummaryByDigestEvicted() *stmtSummaryByDigestEvicted {
+	return &stmtSummaryByDigestEvicted{
+		history: list.New(),
+	}
+}
+
+// spawn a new pointer to stmtSummaryByDigestEvictedElement
+func newStmtSummaryByDigestEvictedElement(beginTime int64, endTime int64) *stmtSummaryByDigestEvictedElement {
+	return &stmtSummaryByDigestEvictedElement{
+		beginTime:    beginTime,
+		endTime:      endTime,
+		digestKeyMap: make(map[string]struct{}),
+	}
+}
+
+// AddEvicted is used add an evicted record to stmtSummaryByDigestEvicted
+func (ssbde *stmtSummaryByDigestEvicted) AddEvicted(evictedKey *stmtSummaryByDigestKey, evictedValue *stmtSummaryByDigest, historySize int) {
+	if evictedValue == nil || evictedValue.history == nil {
+		return
+	}
+
+	evictedValue.Lock()
+	defer evictedValue.Unlock()
+	for e, h := evictedValue.history.Back(), ssbde.history.Back(); e != nil; e = e.Prev() {
+		evictedElement := e.Value.(*stmtSummaryByDigestElement)
+		eBeginTime := evictedElement.beginTime
+		eEndTime := evictedElement.endTime
+
+		// no record in ssbde.history, direct insert
+		if ssbde.history.Len() == 0 && historySize != 0 {
+			record := newStmtSummaryByDigestEvictedElement(eBeginTime, eEndTime)
+			record.addEvicted(evictedKey, evictedElement)
+			ssbde.history.PushFront(record)
+			h = ssbde.history.Back()
+			continue
+		}
+
+		// look for matching history interval
+		// if there are no records in ssbde.history, following code will not be executed. Such situation will probably lead to a bug.
```

Comment:
The comment situation seems never to happen? since line71 already handle this case, can we remove this comment now? It looks scary